File: krisideepweb/src/models/Weather.ts
export interface WeatherData {
  location: string;
  temperature: number;
  humidity: number;
  condition: string;
  conditionHindi: string;
  rainfall: number;
  windSpeed: number;
  timestamp: Date;
  forecast: WeatherForecast[];
}

export interface WeatherForecast {
  date: Date;
  maxTemp: number;
  minTemp: number;
  condition: string;
  conditionHindi: string;
  rainProbability: number;
}

export interface WeatherAlert {
  id: string;
  title: string;
  titleHindi: string;
  description: string;
  descriptionHindi: string;
  severity: string; // low, medium, high, critical
  alertTime: Date;
  expiryTime: Date;
}

export interface WeatherDataJson {
  location: string;
  temperature: number;
  humidity: number;
  condition: string;
  conditionHindi: string;
  rainfall: number;
  windSpeed: number;
  timestamp: string;
  forecast: any[];
}

export interface WeatherForecastJson {
  date: string;
  maxTemp: number;
  minTemp: number;
  condition: string;
  conditionHindi: string;
  rainProbability: number;
}

export interface WeatherAlertJson {
  id: string;
  title: string;
  titleHindi: string;
  description: string;
  descriptionHindi: string;
  severity: string;
  alertTime: string;
  expiryTime: string;
}

export const weatherDataFromJson = (json: WeatherDataJson): WeatherData => ({
  location: json.location,
  temperature: json.temperature,
  humidity: json.humidity,
  condition: json.condition,
  conditionHindi: json.conditionHindi,
  rainfall: json.rainfall,
  windSpeed: json.windSpeed,
  timestamp: new Date(json.timestamp),
  forecast: json.forecast.map((f: any) => weatherForecastFromJson(f)),
});

export const weatherForecastFromJson = (json: WeatherForecastJson): WeatherForecast => ({
  date: new Date(json.date),
  maxTemp: json.maxTemp,
  minTemp: json.minTemp,
  condition: json.condition,
  conditionHindi: json.conditionHindi,
  rainProbability: json.rainProbability,
});

export const weatherAlertFromJson = (json: WeatherAlertJson): WeatherAlert => ({
  id: json.id,
  title: json.title,
  titleHindi: json.titleHindi,
  description: json.description,
  descriptionHindi: json.descriptionHindi,
  severity: json.severity,
  alertTime: new Date(json.alertTime),
  expiryTime: new Date(json.expiryTime),
});

export const weatherDataToJson = (data: WeatherData): WeatherDataJson => ({
  location: data.location,
  temperature: data.temperature,
  humidity: data.humidity,
  condition: data.condition,
  conditionHindi: data.conditionHindi,
  rainfall: data.rainfall,
  windSpeed: data.windSpeed,
  timestamp: data.timestamp.toISOString(),
  forecast: data.forecast.map(weatherForecastToJson),
});

export const weatherForecastToJson = (forecast: WeatherForecast): WeatherForecastJson => ({
  date: forecast.date.toISOString(),
  maxTemp: forecast.maxTemp,
  minTemp: forecast.minTemp,
  condition: forecast.condition,
  conditionHindi: forecast.conditionHindi,
  rainProbability: forecast.rainProbability,
});

export const weatherAlertToJson = (alert: WeatherAlert): WeatherAlertJson => ({
  id: alert.id,
  title: alert.title,
  titleHindi: alert.titleHindi,
  description: alert.description,
  descriptionHindi: alert.descriptionHindi,
  severity: alert.severity,
  alertTime: alert.alertTime.toISOString(),
  expiryTime: alert.expiryTime.toISOString(),
});

File: krishideep_app/lib/models/weather.dart
class WeatherData {
  final String location;
  final double temperature;
  final double humidity;
  final String condition;
  final String conditionHindi;
  final double rainfall;
  final double windSpeed;
  final DateTime timestamp;
  final List<WeatherForecast> forecast;

  WeatherData({
    required this.location,
    required this.temperature,
    required this.humidity,
    required this.condition,
    required this.conditionHindi,
    required this.rainfall,
    required this.windSpeed,
    required this.timestamp,
    required this.forecast,
  });

  Map<String, dynamic> toJson() {
    return {
      'location': location,
      'temperature': temperature,
      'humidity': humidity,
      'condition': condition,
      'conditionHindi': conditionHindi,
      'rainfall': rainfall,
      'windSpeed': windSpeed,
      'timestamp': timestamp.toIso8601String(),
      'forecast': forecast.map((f) => f.toJson()).toList(),
    };
  }

  factory WeatherData.fromJson(Map<String, dynamic> json) {
    return WeatherData(
      location: json['location'],
      temperature: json['temperature'],
      humidity: json['humidity'],
      condition: json['condition'],
      conditionHindi: json['conditionHindi'],
      rainfall: json['rainfall'],
      windSpeed: json['windSpeed'],
      timestamp: json['timestamp'] != null
          ? DateTime.parse(json['timestamp'])
          : DateTime.now(),
      forecast: (json['forecast'] as List)
          .map((f) => WeatherForecast.fromJson(f))
          .toList(),
    );
  }
}

class WeatherForecast {
  final DateTime date;
  final double maxTemp;
  final double minTemp;
  final String condition;
  final String conditionHindi;
  final double rainProbability;

  WeatherForecast({
    required this.date,
    required this.maxTemp,
    required this.minTemp,
    required this.condition,
    required this.conditionHindi,
    required this.rainProbability,
  });

  Map<String, dynamic> toJson() {
    return {
      'date': date.toIso8601String(),
      'maxTemp': maxTemp,
      'minTemp': minTemp,
      'condition': condition,
      'conditionHindi': conditionHindi,
      'rainProbability': rainProbability,
    };
  }

  factory WeatherForecast.fromJson(Map<String, dynamic> json) {
    return WeatherForecast(
      date:
          json['date'] != null ? DateTime.parse(json['date']) : DateTime.now(),
      maxTemp: json['maxTemp'],
      minTemp: json['minTemp'],
      condition: json['condition'],
      conditionHindi: json['conditionHindi'],
      rainProbability: json['rainProbability'],
    );
  }
}

class WeatherAlert {
  final String id;
  final String title;
  final String titleHindi;
  final String description;
  final String descriptionHindi;
  final String severity; // low, medium, high, critical
  final DateTime alertTime;
  final DateTime expiryTime;

  WeatherAlert({
    required this.id,
    required this.title,
    required this.titleHindi,
    required this.description,
    required this.descriptionHindi,
    required this.severity,
    required this.alertTime,
    required this.expiryTime,
  });

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'titleHindi': titleHindi,
      'description': description,
      'descriptionHindi': descriptionHindi,
      'severity': severity,
      'alertTime': alertTime.toIso8601String(),
      'expiryTime': expiryTime.toIso8601String(),
    };
  }

  factory WeatherAlert.fromJson(Map<String, dynamic> json) {
    return WeatherAlert(
      id: json['id'],
      title: json['title'],
      titleHindi: json['titleHindi'],
      description: json['description'],
      descriptionHindi: json['descriptionHindi'],
      severity: json['severity'],
      alertTime: json['alertTime'] != null
          ? DateTime.parse(json['alertTime'])
          : DateTime.now(),
      expiryTime: json['expiryTime'] != null
          ? DateTime.parse(json['expiryTime'])
          : DateTime.now().add(const Duration(hours: 24)),
    );
  }
}
